File: LeetCode/109/Solution.c
#include <stdio.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

//Given a singly linked list where elements are sorted in ascending order, convert it to a height balanced BST.
struct TreeNode* sortedListToBST(struct ListNode* head) 
{
    int len=get_len(head);
    struct ListNode* rootNode=head;
    int i;
    for(i=0;i<(len/2);i++) rootNode=rootNode->next;
    struct TreeNode *right=sortedListToBST(rootNode->next);
	rootNode->next=NULL;
	struct TreeNode *left=sortedListToBST(head);
	TreeNode root;
	root.val=rootNode.val;
	root->left=left;
	root->right=right;
	return root;
}

int get_len(struct ListNode* head)
{	
	if(head==NULL) return 0;
	int len=1;
	while((head=head->next)!=NULL) len++;
	return len;
}

File: LeetCode/109/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        if(head==null) return null;
        int len=getLen(head);
        if(len==1) return new TreeNode(head.val);
        int mid=len/2;
        ListNode cur=head;
        ListNode next=cur.next;
        for(int i=1;i<mid;i++) {
            cur=next;
            next=cur.next;
        }
        TreeNode root=new TreeNode(next.val);
        if(next!=null)root.right=sortedListToBST(next.next);
        else root.right=null;
        cur.next=null;
        root.left=sortedListToBST(head);
        return root;
    }
    
    private int getLen(ListNode head){
        //ListNode cur=head;
        int len=0;
        while(head!=null) {
            head=head.next;
            len++;
        }
        return len;
    }
}